### PR TITLE
Terra UI 375 badge react component

### PR DESCRIPTION
**Pull request by aganapuram**
### Summary

This pull request includes the react component for Terra-Badge. 

@cerner/terra

### CODE REVIEW

File: docs/terra-badge-intent.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-### Label Context
+### Label Intent
```
Comment (bjankord) [inline docs/terra-badge-intent.md]: Replace `Label` with `Badge`

File: src/TerraBadge.jsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
+// For stateless components, use a functional component
+// For stateful components, use a class
+
+// https://facebook.github.io/react/docs/components-and-props.html#functional-and-class-components
```
Comment (bjankord) [inline src/TerraBadge.jsx]: These first 3 comments can be removed.

File: src/TerraBadge.jsx [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+import React, { PropTypes } from 'react';
+import classNames from 'classnames';
+import '../src/terra-badge.scss';
+
+const Badge = (props) => {
+  const classes = classNames(['terra-Badge',
+    { [`terra-Badge--${props.size}`]: props.size },
+     { [`terra-Badge--${props.intent}`]: props.intent },
```
Comment (bjankord) [inline src/TerraBadge.jsx]: nit: Can you align the beginning of this line with the line above?

File: .travis.yml [hunk 4]
```diff
@@ -1,5 +1,7 @@
...
-node_js:
-  - '6'
-  - '6.3'
-script: npm run lint
+script:
+  - yarn test
+cache:
+  yarn: true
+  directories:
+- node_modules
```
Comment (StephenEsser) [inline .travis.yml]: After looking at the failed travis build it looks like this node_modules directory needs tabbed over

File: src/_variables.scss [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+/*$terra-label-badge-border-radius: 1000px !default; //really large number foundation uses 1000*/
+$terra-badge-child-margin: 0.357em !default;
+
+$terra-badge-border-radius: 0.25em !default;
+$terra-badge-display: inline-block !default;
+
+$terra-badge-font-weight: 700 !default;
+$terra-badge-line-height: 1 !default;
+
+$terra-badge-padding-bottom: 0.5rem !default;
```
Comment (aganapuram) [inline src/_variables.scss]: Increased the padding on top and bottom to provide proper spacing for the icon in the badge.

File: src/TerraBadge.jsx [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  ]);
+
+  const text = props.text ? <span className="terra-Badge-text">{props.text}</span> : null;
+
+  const order = props.isReversed ?
+    [text, props.icon, props.children] :
+    [props.icon, text, props.children];
+
+  const attributes = { className: classes, htmlFor: text, 'aria-label': text, tabIndex: -1 };
+  return React.createElement('label', attributes, ...order);
```
Comment (bjankord) [inline src/TerraBadge.jsx]: What are your thoughts on making this a `span` element instead of a `label` element?

File: src/_variables.scss [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  medium: 1rem,
+  large: 1.4rem,
+  huge: 1.7rem
+) !default;
+
+/*$terra-label-badge-border-radius: 1000px !default; //really large number foundation uses 1000*/
+$terra-badge-child-margin: 0.357em !default;
+
+$terra-badge-border-radius: 0.25em !default;
+$terra-badge-display: inline-block !default;
```
Comment (mhemesath) [inline src/_variables.scss]: I don't think this should be a variable

Comment (mhemesath) [inline src/_variables.scss]: In general I don't think many of these things should be themeable until its explicitly requested.

File: src/terra-badge.scss [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  padding-right: $terra-badge-padding-left;
+  padding-top: $terra-badge-padding-top;
+  position: relative;
+  text-align: center;
+  text-decoration: none;
+  text-transform: $terra-button-text-transform;
+  white-space: nowrap;
+
+  &:last-child {
+    @include terra-margin-end(0);
```
Comment (mhemesath) [inline src/terra-badge.scss]: `.terra-Badge` has no margin so why is this needed?

File: src/_variables.scss [hunk 9]
```diff
@@ -0,0 +1,58 @@
+$terra-badge-font-sizes: (
+  tiny: 0.5rem,
+  small: 0.8rem,
+  medium: 1rem,
+  large: 1.4rem,
+  huge: 1.7rem
+) !default;
+
+/*$terra-label-badge-border-radius: 1000px !default; //really large number foundation uses 1000*/
```
Comment (bjankord) [inline src/_variables.scss]: Looks like this line of commented out code can be removed.

File: src/TerraBadge.jsx (hunk 6 above)
Comment (mhemesath) [inline src/TerraBadge.jsx]: I think a span fits better here. Label doesn't fit because labels are associated to controls... and badges are not.

File: docs/terra-badge-order.md [hunk 10]
```diff
@@ -0,0 +1,4 @@
+### React Badge order
+| Prop            | Default   | Default   | Description                                                          											   |
```
Comment (amf9t2) [inline docs/terra-badge-order.md]: Default => Type ?  for Boolean.

File: src/TerraBadge.jsx [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+import React, { PropTypes } from 'react';
+import classNames from 'classnames';
+import '../src/terra-badge.scss';
+
+const Badge = (props) => {
+  const classes = classNames(['terra-Badge',
+    { [`terra-Badge--${props.size}`]: props.size },
+     { [`terra-Badge--${props.intent}`]: props.intent },
+  ]);
```
Comment (StephenEsser) [inline src/TerraBadge.jsx]: You don't need classNames for this, classNames is used to conditionally add classes together. So if the condition is true it will add the class, when the condition is false the class will be omitted. 

The conditions you are using above are props.size, and props.intent, which will always be true because they have a default. These lines are saying if props.size exists add this class, but props.size will always exist.

The classes can just be defined inline or assign into a local variable without the classNames import. 

```
const classes = `terra-Badge terra-Badge--${props.size} terra-Badge--${props.intent}`;
```

File: src/TerraBadge.jsx [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+  const text = props.text ? <span className="terra-Badge-text">{props.text}</span> : null;
+
+  const order = props.isReversed ?
+    [text, props.icon, props.children] :
+    [props.icon, text, props.children];
```
Comment (StephenEsser) [inline src/TerraBadge.jsx]: You can make this ternary a single line. The Button was limited to 100 characters on a line so it was wrapped, but we decided to removed the 100 character limit per line in the eslintrc which is already in your project.

File: src/TerraBadge.jsx (hunk 6 above)
Comment (StephenEsser) [inline src/TerraBadge.jsx]: +1

**Comment by StephenEsser** [comment]
You will need to run `yarn install` to generate a yarn.lock file and check that in before the travis build will success

File: stories/index.js [hunk 13]
```diff
@@ -0,0 +1,74 @@
+import React from 'react';
+import { storiesOf, action, linkTo } from '@kadira/storybook';
+import Badge from '../src/TerraBadge';
+
+storiesOf('Badge', module)
+  .add('Default', () => (
+    <Badge />
+    ));
```
Comment (StephenEsser) [inline stories/index.js]: This default badge doesn't display in the Storybook because the :empty css sets the display to none. I'd either remove it, add 'default' text, or update the description to indicate it would be hidden because it was empty

File: docs/terra-badge-intent.md (hunk 1 above)
Comment (aganapuram) [inline docs/terra-badge-intent.md]: @bjankord  replaced Label with Badge

File: docs/terra-badge-order.md (hunk 10 above)
Comment (aganapuram) [inline docs/terra-badge-order.md]: @amf9t2 changed the column name to Type.

File: src/TerraBadge.jsx (hunk 2 above)
Comment (aganapuram) [inline src/TerraBadge.jsx]: @bjankord removed the comments.

File: src/TerraBadge.jsx (hunk 3 above)
Comment (aganapuram) [inline src/TerraBadge.jsx]: @bjankord fixed the alignment.

File: src/TerraBadge.jsx (hunk 11 above)
Comment (aganapuram) [inline src/TerraBadge.jsx]: @StephenEsser thank you, we didn't know about this. We have made the necessary changes.

File: src/TerraBadge.jsx (hunk 12 above)
Comment (aganapuram) [inline src/TerraBadge.jsx]: @StephenEsser thank you, made the ternary a single line.

File: src/TerraBadge.jsx (hunk 6 above)
Comment (aganapuram) [inline src/TerraBadge.jsx]: @bjankord @mhemesath @StephenEsser  thank you, span makes more sense. We have updated it.

File: src/terra-badge.scss (hunk 8 above)
Comment (aganapuram) [inline src/terra-badge.scss]: @mhemesath yes, it is not needed. Removed it.

File: stories/index.js (hunk 13 above)
Comment (aganapuram) [inline stories/index.js]: @StephenEsser thank you, updated the storybook with a description.

File: src/TerraBadge.jsx [hunk 14]
```diff
@@ -0,0 +1,29 @@
+import React, { PropTypes } from 'react';
```
Comment (bjankord) [inline src/TerraBadge.jsx]: Please rename this file to `Badge.jsx`

File: tests/TerraBadge.test.jsx [hunk 15]
```diff
@@ -0,0 +1,128 @@
+import React from 'react';
```
Comment (bjankord) [inline tests/TerraBadge.test.jsx]: Please rename this to `Badge.test.jsx`

File: src/TerraBadge.jsx (hunk 14 above)
Comment (aganapuram) [inline src/TerraBadge.jsx]: @bjankord renamed the file to Badge.jsx.

File: tests/TerraBadge.test.jsx (hunk 15 above)
Comment (aganapuram) [inline tests/TerraBadge.test.jsx]: @bjankord renamed the file to Badge.test.jsx

File: .stylelintrc [hunk 16]
```diff
@@ -1,7 +1,7 @@
 {
   "extends": "stylelint-config-sass-guidelines",
   "rules": {
-    "max-nesting-depth": 2,
+    "max-nesting-depth": 3,
```
Comment (mhemesath) [inline .stylelintrc]: does this need to be 3? I don't see any selectors that violate that

Comment (aganapuram) [inline .stylelintrc]: @mhemesath this file was overwritten by the generator. Should we change it back to 2?

Comment (mhemesath) [inline .stylelintrc]: oh gotcha. You can leave it at 2, thats fine

**Comment by mhemesath** [comment]
+1

File: .stylelintrc (hunk 16 above)
Comment (aganapuram) [inline .stylelintrc]: changed it back to 2.

**Comment by bjankord** [comment]
@aganapuram Can you run `yarn build` on the project. This will generate the `lib` directory with `TerraBadge.js`.

File: docs/terra-badge-intent.md [hunk 17]
```diff
@@ -0,0 +1,13 @@
+### Badge Intent
+
+Badges can be colored by placing any of the following classes on the `terra-Badge` component:
+
+| Class Name                | Behavior                                                                 |
+|---------------------------|--------------------------------------------------------------------------|
+| `terra-Badge--default`   | Sets the colors of the badge based on the default badge color scheme      |
```
Comment (mhemesath) [inline docs/terra-badge-intent.md]: i thought we weren't documenting class names?

Comment (aganapuram) [inline docs/terra-badge-intent.md]: Can we replace the class names with the corresponding string values?

**Comment by aganapuram** [comment]
@mhemesath  updated the docs with prop values instead of class names.

File: src/Badge.jsx [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+  return React.createElement('span', { className: classes }, ...order);
+};
+
+Badge.propTypes = {
+  text: PropTypes.string,
+  size: PropTypes.oneOf(['tiny', 'small', 'medium', 'large', 'huge']),
+  intent: PropTypes.oneOf(['default', 'primary', 'secondary', 'info', 'warning', 'positive', 'negative']),
+  isReversed: PropTypes.bool,
+  icon: PropTypes.element,
+  children: React.PropTypes.node,
```
Comment (zhongyn) [inline src/Badge.jsx]: 'React.' can be removed here

File: .storybook/head.html [hunk 19]
```diff
@@ -0,0 +1,6 @@
+<!-- Set terra-Application on the html tag -->
+<script>
+  var root = document.getElementsByTagName( 'html' )[0];
+  root.setAttribute( 'class', 'terra-Application' );
+  root.setAttribute( 'dir', 'ltr' );
+</script>
```
Comment (zhongyn) [inline .storybook/head.html]: the generator has been updated: https://github.com/cerner/generator-terra-module/blob/master/generators/app/templates/.storybook/head.html, need to run it again

**Comment by jpruetting** [comment]
**[Storybook Hub](https://hub.getstorybook.io/)** is building storybooks for every commit in this PR.

**Latest Storybook: https://62803838-9db0-4955-b28b-bea68f7fbbd2.sbook.io/**
<sub>( Visit [here](https://hub.getstorybook.io/app/github.com/cerner/terra-badge/branch/TerraUI375_BadgeReactComponent/) for more info )</sub>

> _This comment is automatically added by Storybook Hub_

**Comment by sid141288** [comment]
@zhongyn ran the latest generator and updated the code.

File: src/Badge.jsx [hunk 20]
```diff
@@ -0,0 +1,29 @@
+import React, { PropTypes } from 'react';
+import '../src/terra-badge.scss';
+
+const Badge = (props) => {
```
Comment (mjhenkes) [inline src/Badge.jsx]: You're not yet supporting custom props. In the interest of not dragging out this PR. I've logged an issue against badge that can be addressed later.
https://github.com/cerner/terra-badge/issues/8

**Comment by windse7en** [comment]
We can remove all `*.html` files in `tests/visual` and the `ejs` file. Those files were used to test previous OSS scss code. We will update to write React code for those examples.

**Comment by zhongyn** [comment]
+1

**Comment by sid141288** [comment]
@windse7en @mjhenkes Updated to remove the unnecessary html files and added custom props.
